Refactor Compile/Optcompile

Drup opened the pull request:
I always felt that the `Compile` and `Optcompile` modules were a slight violation of the DRY principle, so this patchset ~removes them~ fixes them.

This patchset factorize the common content of `Compile` and `Optcompile` into a third module, `Libcompile`. It also make the whole pipeline significantly more modular and, I hope, easier to tinker with.

I originally made this patch for the eliom fork, where each file is first typechecked, split and untyped into two parsetrees and then typechecked again and compiled. Given the current pipeline code, this was impossible to write reasonably.
With this patchset, it's just a matter of plumbing: https://github.com/ocsigen/ocaml-eliom/blob/master/driver/compile.ml#L62-L76

I believe this could be useful in a wider context, to easily build custom pipelines or tinker with the existing one. For example to test untypedast, I made a compiler which types, untypes and retypes. 
Finally, it also avoids a big chunk of code duplication.

The patches are not exactly atomic. I'll squash when the time arrives.

I'm not exactly sure how to test this. At least it bootstraps.

Drup committed:
Add compile_common.ml which contains the basic compilation pipeline.

Factorize the part from compile.ml and optcompile.ml.

nojb commented on Changes:
```diff
@@ -350,6 +350,10 @@ Working version
   (Hugo Heuzard, review by Mark Shinwell)
 
+- GPR#__ : Add the module Libcompile, which factorize the common
+  part in Compile and Optcompile. This also make the pipeline more modular.
```
s/factorize/factorizes/
s/make/makes/

Drup replied:
Fixed, thanks!

mshinwell commented:
I will read this.

mshinwell commented:
I had a look at this using patdiff, which is excellent for checking this sort of patch, and didn't notice anything semantically wrong.  However I'm only reasonably confident about that.  One thing which might simplify matters further is to make the libcompile code into a functor (I think this will eliminate more duplication).  It also avoids the kind of "wrap" notion which always seems slightly dubious.

I would maybe think of another name rather than "libcompile" as it's not really a separate library.

A couple of stylistic points: I think blank lines in functions should be avoided (if the function is unclear or too long, refactor it in another way) and spaces before record field semicolons are unusual.  Also, it would be nice if all of the functions (etc.) in the libcompile interface were documented.

This patch would seem to me to be a very good step along the way to making ocamlopt emit bytecode as well as native code.  Maybe that's worth trying?  In particular you could apply the same kind of simplifications as you have here to `main.ml` and `optmain.ml` -- there are various unnecessary inconsistencies there at present (e.g. `unset` versus `clear`).  If you merge all the options together, with bytecode-only or native-only ones annotated for the `-help` output, then I suspect there isn't much more work to do in order to reach that goal.  I believe @damiendoligez is a particular fan of achieving that.

Drup commented:
Thanks for the review! Indeed, the diff is quite hard to read (in fact, I had to port this from 4.03, and I ended up partially rewriting it because the diffs were impossible to work with).

What about renaming `Compile` to `Bytecompile`, and making this new module `Compile` ? It's not very consistent with the rest of the driver module but ....

I will look into turning this into a functor, but I'm not so convinced. The pipeline after typing differs significantly between byte and opt. Unless we standardize over a common API, I don't see how we can improve the situation further. I also would prefer to keep the "combinator" approach used here, as I think it's more flexible.

I completely agree with your last point about merging the byte and opt driver, this is the end goal. 
I remember asking @chambart about having the opt compiler emit bytecode. I don't remember the exact answer, but it was not positive. In any cases, all this is probably better left for the next PRs!

mshinwell commented:
The next step should probably be to do some large-scale testing of this on OPAM.  Maybe talk to the OCaml Labs guys to see if they can assist with setting up a build?

Drup commented:
After taking another look, I'm really not sure how to improve it by using a functor. We could define a functor just for the `wrap` function, but I don't really see that as an improvement. The flexibility offered by the combinator version is essential for defining non-orthodox pipelines. Could you clarify what you had in mind?

I fixed your stylistic remarks and added some more documentation comments.

I asked @avsm and @AltGr. They said they have some testing platform that should be available soon to test compilers, so I'll wait for that.

Drup commented:
@dra27 Is the appveyor failure relevant? I can't find what's wrong in the log. If not, can you relaunch the job?

Drup commented:
Thanks to @kit-ty-kate, the output of the opam run is [here](http://51.145.134.72:8080/). Apparently, there is an issue with annot files.

Drup commented:
@shindere I would like to write a test that tries to compile a file with ocamlc and ocamlopt, and in each case verifies that `.annot` files have been generated. I have tried to look at similar things in the testsuite but without success, and I don't really understand the DSL used in test files. How do I do something like that ?

shindere commented:
Gabriel Radanne (2018/04/12 14:23 +0000):
> @shindere I would like to write a test that tries to compile a file
> with ocamlc and ocamlopt, and in each case verifies that `.annot`
> files have been generated. I have tried to look at similar things in
> the testsuite but without success, and I don't really understand the
> DSL used in test files. How do I do something like that ?

Yeah I'm sorry aobut the lack of documentation of the DSL and ocamltest
tool in general.

You could do something like this (I'm replying in a mail, I hope GH will
preserve the presentation!):

```
(* TEST

flags = "..." (* flags common to both compilers, if any *)
ocamlc_flags = "..." (*ocamlc-specificflags *)
ocamlopt_flags = "..." (* ocamlopt-specific flags,if any *)
script = "sh ${test_source_directory}/check-annot"

* setup-ocamlc.byte-build-env
** ocamlc.byte
*** script

(* same for ocamlopt.byte *)
*)
```

And create the check-annot script like this:

```
if [ -f ${test_build_directory}/myfile.annot ]; then
  exit ${TEST_PASS}
else
  exit ${TEST_FAIL}
fi
```

Let me know if you encounter issues. I can definitely review the test.

Drup commented:
@shindere Thanks! That's very helpful. I have something that works in the case the compiler succeed, but I would also like to allow the compiler to fail. How do I do that?

Drup committed:
Test the presence of annot files after compilation.

shindere commented:
Gabriel Radanne (2018/04/12 15:39 +0000):
> @shindere Thanks! That's very helpful. I have something that works in
> the case the compiler succeed, but I would also like to allow the
> compiler to fail. How do I do that?

Oh sorry, I think I forgot to repsond to this.

ocamlc_byte_exit_status = "2"

Drup commented:
@shindere Perfect. It works nicely now.

shindere commented:
Gabriel Radanne (2018/04/12 16:16 +0000):
> @shindere Perfect. It works nicely now.

Great!!

Drup commented:
So, the new driver code didn't emit `.annot` files when a compilation was successful. This is now fixed and I added a test that checks that `.annot` files are emitted if a compilation is a success, a failure or if `-i` is provided. The [new opam run](http://51.145.134.72:8080/) is identical to trunk and travis is happy.
For some reason, the "typeonly" test fails on appveyor. I have no idea why.

damiendoligez commented:
> For some reason, the "typeonly" test fails on appveyor. I have no idea why.

I'm investigating this.

damiendoligez commented:
It fails because of a subtle interaction between `-i`, `-annot`, and the `-o` added by ocamltest, and the fact that executable files have an extension under Windows and not under Unix.

The solution (credit @shindere) is to add a line `compile_only = "true"` right after the `flags = ...` line in the test file, to prevent ocamltest from adding the `-o typeonly.byte.exe` option.

Drup commented:
@damiendoligez and @shindere Thanks! I added the appropriate line in `typeonly.ml` and rebased. Let's see the results.

Drup commented:
Tests are finally passing. @damiendoligez Are you merging as-is?

shindere commented:
@Drup: if you don't mind, I would suggest to rebase this on latest trunk
and, more importantly, to cleanup the history a bit. For instance, I
think the two most recent commits could be squashed. Also, ideally, I
think when a commit adds a file, like libcompile.ml, that very same
commit should update the dependencies so that they are as correct as
possible for every commit. This turns out to be important when
bisecting.

Drup commented:
It's already rebased. The multiple commits are for review purposes. I'll squash when it's approved.

shindere commented:
Gabriel Radanne (2018/05/29 02:36 -0700):
> It's already rebased. The multiple commits are for review purposes.
> I'll squash when it's approved.

Okay thanks!

mshinwell commented:
@Drup I tried to read the rest of this.  I couldn't find a separate changeset which fixed the stylistic/documentation issues you said you had attended to.  Do you know what's happened there?

Some other remarks:
- `#!/usr/bin/sh` at the top of the annot-checking script seems unusual/wrong.  Also, I recommend always having "`set -eu`" (and probably "`set -eu -o pipefail`") at the top of every shell script, no matter how short it may be at the moment.

- There is one unsightly "`end ;`" instead of "`end;" in `libcompile.ml`.

Please ensure this isn't squashed (or similar) before merging as it will be difficult to review.

shindere commented:
Mark Shinwell (2018/05/30 07:06 -0700):
> Some other remarks:
> - `#!/usr/bin/sh` at the top of the annot-checking script seems
> unusual/wrong.  Also, I recommend always having "`set -eu`" (and
> probably "`set -eu -o pipefail`") at the top of every shell script, no
> matter how short it may be at the moment.

pipefail is not Posix so wouldn't work with dash whichis used e.g. on
Travis, I think.

Drup commented:
@mshinwell I was thinking about the documentation aspects, which are done in aa3963d . I forgot about the coding style part. I'll do that later today.

I copied the file from @shindere. so I'll let you two debate how it should look like. :)

shindere commented:
Gabriel Radanne (2018/05/30 07:35 -0700):
> I copied the file from @shindere.

Oops! :) Which file did you take inspiration from, please?

Drup commented:
@shindere Your answer from the [the 12th of April](https://github.com/ocaml/ocaml/pull/1703#issuecomment-380834346) on this thread.

shindere commented:
Ah, so it's not a testsuite file, okay.

I guess I didn't expect this answer to be taken that literally. It
should be #/bin/sh, of course.

Drup commented:
I just rebased, fixed the `sh` path and the coding style.

I would appreciate if this was merged, as the rebasing of this is not particularly pleasant. If documentation is not sufficient, I can fix it in a later patch.

damiendoligez commented on driver/libcompile.mli:
```diff
@@ -0,0 +1,91 @@
+(*                                                                        *)
+(*                            Gabriel Radanne                             *)
+(*                                                                        *)
+(*   Copyright 1996 Institut National de Recherche en Informatique et     *)
```
You should change this to the correct year and institution.

damiendoligez reviewed (commented):
I haven't reviewed in detail. @mshinwell are you available to do it soon?

mshinwell commented:
Yes, I should be able to deal with this at the start of next week.

dra27 commented on driver/compile.mli:
```diff
@@ -13,9 +13,23 @@
-val interface: formatter -> string -> string -> unit
-val implementation: formatter -> string -> string -> unit
+val interface: formatter -> sourcefile:string -> outputprefix:string -> unit
+val implementation: formatter -> sourcefile:string -> outputprefix:string -> unit
```
This line's more than 80 chars.

dra27 commented on testsuite/tests/tool-ocaml-annot/success.ml:
```diff
@@ -0,0 +1,17 @@
+*** script
+*)
+
+(* Check that .annot files are emitted in case of regular succesful compilation. *)
```
This line's more than 80 chars

dra27 reviewed (dismissed):
Sorry for the nits - when #1148 is merged, check-typo "violations" become CI errors. There's also white-space at the end of lines in `driver/compile.mli`, `driver/libcompile.mli` and `driver/optcompile.mli`

Drup commented:
@damiendoligez @dra27 Done.

dra27 commented:
Thanks @Drup! I dismissed the review rather than "approving" it, as that would have looked a bit weird...

Drup commented:
4.07 is released, the world cup is finished. @mshinwell, can I get a review? :)

Drup commented:
@mshinwell one down, one to go ? :)

mshinwell commented on driver/libcompile.mli:
```diff
@@ -0,0 +1,90 @@
+val emit_signature : info -> Parsetree.signature -> Typedtree.signature -> unit
+(** [emit_signature info parsetree typedtree] emits the [.cmi] file
+    containing the signature.
+*)
```
"the signature" -> "the given signature"

mshinwell commented on driver/libcompile.mli:
```diff
@@ -0,0 +1,90 @@
+
+val print_if :
+  info -> bool ref -> (Format.formatter -> 'a -> unit) -> 'a -> 'a
+(** [print_if i flag fmt x] prints [x] with [fmt] on [i] if [b] is true. *)
```
This should move elsewhere, maybe to `Misc`.

mshinwell commented on driver/libcompile.mli:
```diff
@@ -0,0 +1,90 @@
+(**************************************************************************)
```
I still think the name of this module isn't clear.  What about `Compile_common` or similar?

mshinwell commented on driver/libcompile.mli:
```diff
@@ -0,0 +1,90 @@
+val cmx : info -> string
+val obj : info -> string
+val annot : info -> string
+(** Return the filename of some compiler products associated
```
"products" -> "build artifacts"

mshinwell commented on driver/libcompile.mli:
```diff
@@ -0,0 +1,90 @@
+val typecheck_impl :
+  info -> Parsetree.structure -> Typedtree.structure * Typedtree.module_coercion
+(** [typecheck_impl info parsetree] typechecks an implementation and returns
+    the typedtree of the associated module, along with a coercion again
```
"again" -> "against"

mshinwell commented:
Can we please run this against OPAM once again after addressing the final review comments and rebasing?

Drup commented:
@mshinwell I addressed all your comments.

Rebasing after #1913 was extremely painful and I was more or less forced to squash everything. I do not want to have to rebase this patchset again. Whatever the opam repository has to say, it'll be minor and I can fix it afterwards. I insist that we merge as soon as CI passes.

mshinwell commented:
OK, that's fair enough.  Let's get an OPAM run going at that point though, in case there is something we've missed.

mshinwell commented:
ah, and Changes needs the new entry formatting properly.

Drup commented:
> ah, and Changes needs the new entry formatting properly.

Done.

v-gb commented:
@drup sorry, I've had my dose of conflicts too. Would you please rename the field info.ppf as info.ppf_dump?

Drup commented:
> @Drup sorry, I've had my dose of conflicts too. Would you please rename the field info.ppf as info.ppf_dump?

Done.

Drup commented:
Everything is green!

Drup commented:
:tada:

v-gb commented on driver/optcompile.ml:
```diff
@@ -13,136 +13,74 @@
-      ~always:(fun () -> Stypes.dump (Some (outputprefix ^ ".annot")))
-  ))
+let implementation ~backend ~sourcefile ~outputprefix =
+  Compmisc.with_ppf_dump ~fileprefix:(outputprefix ^ ".cmo") @@ fun ppf_dump ->
```
This should be .cmx, not .cmo.

v-gb commented on driver/compile_common.ml:
```diff
@@ -0,0 +1,118 @@
+let cmo i = i.outputprefix ^ ".cmo"
+let annot i = i.outputprefix ^ ".annot"
+
+let init ppf_dump ~init_path ~tool_name ~sourcefile ~outputprefix =
```
~init_path is a strange name. Shouldn't this be called ~native?

Drup replied:
Well, it's the boolean that control if paths should be initialized (which is apparently needed for native). I wouldn't object renaming it.

alainfrisch commented:
Note: this PR introduced a regression reported in [MPR#7918](https://caml.inria.fr/mantis/view.php?id=7918).

gasche commented on driver/compile_common.ml:
```diff
@@ -0,0 +1,118 @@
+      )
+    ~exceptionally:(fun () ->
+        Misc.remove_file (obj info);
+        Misc.remove_file (cmx info);
```
The regression in [MPR#7918](https://caml.inria.fr/mantis/view.php?id=7918) was introduced here: this `exceptionnally` catches `Warnings.check_fatal` errors, but it is written with ocamlopt in mind, and does not remove the `cmo` file.

(The code in trunk has changed and `wrap_compilation` is gone, I haven't looked at how to fix it yet.)

gasche replied:
(Fix proposed. It's easier with the more regular code structure in trunk.)

gasche commented:
I propose a fix for the `check_fatal` regression in https://github.com/ocaml/ocaml/pull/2257.
